server/src/main/java/org/elasticsearch/common/time/Iso8601Parser.java

```diff
@@ -0,0 +1,521 @@
+
+        if (pos == 20) return Result.error(20);   // didn't find a number at all
+
+        // multiply it by the correct multiplicand to get the nanos
+        nanos *= NANO_MULTIPLICANDS[29 - pos];
+
+        if (len == pos) {
+            return new Result(withZoneOffset(years, months, days, hours, minutes, seconds, nanos, defaultTimezone));
+        }
+        if (isZoneId(str, pos)) {
+            ZoneId timezone = parseZoneId(str, pos);
+            return timezone != null
+                ? new Result(withZoneOffset(years, months, days, hours, minutes, seconds, nanos, timezone))
+                : Result.error(pos);
+        }
+
+        // still chars left at the end - string is not valid
+        return Result.error(pos);
+    }
+
+    private static boolean isZoneId(CharSequence str, int pos) {
+        // all region zoneIds must start with [A-Za-z] (see ZoneId#of)
+        // this also covers Z and UT/UTC/GMT zone variants
+        char c = str.charAt(pos);
+        return c == '+' || c == '-' || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
+    }
+
+    /**
+     * This parses the zone offset, which is of the format accepted by {@link java.time.ZoneId#of(String)}.
+     * It has fast paths for numerical offsets, but falls back on {@code ZoneId.of} for non-trivial zone ids.
+     */
+    private ZoneId parseZoneId(CharSequence str, int pos) {
+        int len = str.length();
+        char first = str.charAt(pos);
+
+        if (first == 'Z' && len == pos + 1) {
+            return ZoneOffset.UTC;
+        }
+
+        boolean positive;
+        switch (first) {
+            case '+' -> positive = true;
+            case '-' -> positive = false;
+            default -> {
+                // non-trivial zone offset, fallback on the built-in java zoneid parser
+                try {
+                    return ZoneId.of(str.subSequence(pos, str.length()).toString());
+                } catch (DateTimeException e) {
+                    return null;
+                }
+            }
+        }
+        pos++;  // read the + or -
+
+        Integer hours = parseInt(str, pos, pos += 2);
+        if (hours == null) return null;
+        if (len == pos) return ofHoursMinutesSeconds(hours, 0, 0, positive);
+
+        boolean hasColon = false;
+        if (str.charAt(pos) == ':') {
+            pos++;
+            hasColon = true;
+        }
+
+        Integer minutes = parseInt(str, pos, pos += 2);
+        if (minutes == null) return null;
+        if (len == pos) return ofHoursMinutesSeconds(hours, minutes, 0, positive);
+
+        // either both dividers have a colon, or neither do
+        if ((str.charAt(pos) == ':') != hasColon) return null;
+        if (hasColon) {
+            pos++;
+        }
+
+        Integer seconds = parseInt(str, pos, pos += 2);
+        if (seconds == null) return null;
+        if (len == pos) return ofHoursMinutesSeconds(hours, minutes, seconds, positive);
+
+        // there's some text left over...
```

Comment:
is it  be worth rephrasing  to? `        // there's some text left over... will return error`